src/python/pants/backend/terraform/tool.py

```diff
@@ -2,44 +2,165 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 from __future__ import annotations
 
+import json
+import logging
+import os
 from dataclasses import dataclass
 
-from pants.core.util_rules.external_tool import (
-    DownloadedExternalTool,
-    ExternalToolRequest,
-    TemplatedExternalTool,
+from packaging.specifiers import InvalidSpecifier, SpecifierSet
+from packaging.version import InvalidVersion, Version
+
+from pants.engine.environment import Environment, EnvironmentRequest
+from pants.engine.fs import EMPTY_DIGEST, Digest
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import (
+    BinaryNotFoundError,
+    BinaryPath,
+    BinaryPathRequest,
+    BinaryPaths,
+    BinaryPathTest,
+    Process,
+    ProcessCacheScope,
+    ProcessResult,
 )
-from pants.engine.fs import EMPTY_DIGEST, Digest, MergeDigests
-from pants.engine.internals.selectors import Get
-from pants.engine.platform import Platform
-from pants.engine.process import Process
 from pants.engine.rules import collect_rules, rule
+from pants.option.subsystem import Subsystem
 from pants.util.logging import LogLevel
-from pants.util.meta import classproperty
+from pants.util.ordered_set import OrderedSet
+
+logger = logging.getLogger(__name__)
 
 
-class TerraformTool(TemplatedExternalTool):
-    options_scope = "download-terraform"
-    name = "terraform"
+class TerraformSubsystem(Subsystem):
+    options_scope = "terraform"
     help = "Terraform (https://terraform.io)"
 
-    default_version = "1.0.7"
-    default_url_template = (
-        "https://releases.hashicorp.com/terraform/{version}/terraform_{version}_{platform}.zip"
```

Comment:
So, even if we support using this from the PATH, it seems like we should still keep the support for fetching it? It certainly makes it easier for us as consumers to be able to run hermetically from an empty path.

Reply:
I argue in https://github.com/pantsbuild/pants/pull/12955#issuecomment-929283138 why I think we should "keep it simple"

Reply:
cc @benjyw and @kaos, I don't think we've heard your opinion yet on whether we should support for Terraform 1) only use from local PATH, 2) only use from Pants downloading, or 3) both.

This decision also will apply to Go, which for now only supports local PATH. When we add "immutable caches" so that Pants downloading is performant, we'll need to decide if we still keep local PATH support or not.

Reply:
I think that it makes contributors lives easier to bootstrap this for you, and that in consumer respositories there will be users who might "use the repository" without actually executing terraform directly for sideeffects. Basically: the convenience provided here for our contributors can be mirrored in user repositories as well.

So my vote would in general be to support both bootstrapping and using a system install (with a `--system` toggle) where possible. It seems like we can't for `go` (yet!) for practical reasons, but this might apply there as well in future.

Reply:
Okay, that's sensible enough to me. 

We may need to tweak the options interface for `ExternalTool` to work better with installing from PATH. For example, `--version` might not as useful, given Tom's recommendation that we should support semantic versions. All the users of this pattern (Java, Go, Terraform) are still experimental, so we can punt on that redesign - but it'd be good to open a ticket and figure out before we stabilize these backends.

Reply:
So is the consensus to use ExternalTool's option plus these options (using `terraform` as an example): `--terraform-system`, `--terraform-search-paths`, and `--terraform-version-constraint` ?

Reply:
In my comment above, I speculate that we probably want to rename the `ExternalTool` options if system installs are supported. Rather than `--terraform-version` (downloads) and `--terraform-version-constraint` (system), we probably want something like `--terraform-download-version` and `--terraform-system-version-constraint`.

But I think we can punt on redesigning the option names, so long as we have a ticket tracking it.